src/main/java/net/minecraftforge/datafix/ForgeSchema.java

```diff
@@ -0,0 +1,340 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.datafix;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.mojang.datafixers.DSL;
+import com.mojang.datafixers.DataFixUtils;
+import com.mojang.datafixers.schemas.Schema;
+import com.mojang.datafixers.types.Type;
+import com.mojang.datafixers.types.families.RecursiveTypeFamily;
+import com.mojang.datafixers.types.families.TypeFamily;
+import com.mojang.datafixers.types.templates.RecursivePoint;
+import com.mojang.datafixers.types.templates.TaggedChoice;
+import com.mojang.datafixers.types.templates.TypeTemplate;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * A partial carbon copy of the {@link Schema} class.
+ * But this allows us to delay the setup of the type hierarchy until we need it.
+ */
+class ForgeSchema extends Schema
+{
+    //Can not be final, needs to be initialized on the fly!
+    //Constructor invocation order and the what not!
+    private Map<String, Supplier<TypeTemplate>> TYPE_TEMPLATES = Maps.newConcurrentMap();
```

Comment:
I believe `TYPE_TEMPLATES` can be a normal hash map. It is in `Schema` and in `registerType` it is set to a hash map. If not then the normal hash map variant in `registerType` should be changed to be concurrent as well.

Reply:
Fixed that in already. It needs to be concurrent due to the concurrent firing of the event.